// @mantine/charts/cjs/RadialBarChart/RadialBarChart.cjs
'use client';
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var react = require('react');
var recharts = require('recharts');
var core = require('@mantine/core');
var ChartLegend = require('../ChartLegend/ChartLegend.cjs');
var RadialBarChart_module = require('./RadialBarChart.module.css.cjs');

const defaultProps = {
  barSize: 20,
  startAngle: 90,
  endAngle: -270,
  withBackground: true,
  withTooltip: true
};
const varsResolver = core.createVarsResolver(
  (theme, { emptyBackgroundColor }) => ({
    root: {
      "--chart-empty-background": emptyBackgroundColor ? core.getThemeColor(emptyBackgroundColor, theme) : void 0
    }
  })
);
const RadialBarChart = core.factory((_props, ref) => {
  const props = core.useProps("RadialBarChart", defaultProps, _props);
  const {
    classNames,
    className,
    style,
    styles,
    unstyled,
    vars,
    data,
    barSize,
    withBackground,
    dataKey,
    radialBarProps,
    radialBarChartProps,
    withLabels,
    withLegend,
    legendProps,
    withTooltip,
    tooltipProps,
    startAngle,
    endAngle,
    ...others
  } = props;
  const [highlightedArea, setHighlightedArea] = react.useState(null);
  const getStyles = core.useStyles({
    name: "RadialBarChart",
    classes: RadialBarChart_module,
    props,
    className,
    style,
    classNames,
    styles,
    unstyled,
    vars,
    varsResolver
  });
  const theme = core.useMantineTheme();
  const dataWithResolvedColor = data.map(({ color, ...item }) => {
    const resolvedColor = core.getThemeColor(color, theme);
    return {
      ...item,
      fill: resolvedColor,
      fillOpacity: highlightedArea ? highlightedArea === item.name ? item.opacity || 1 : 0.05 : item.opacity || 1
    };
  });
  const { resolvedClassNames, resolvedStyles } = core.useResolvedStylesApi({
    classNames,
    styles,
    props
  });
  return /* @__PURE__ */ jsxRuntime.jsx(core.Box, { ref, ...getStyles("root"), ...others, children: /* @__PURE__ */ jsxRuntime.jsx(recharts.ResponsiveContainer, { children: /* @__PURE__ */ jsxRuntime.jsxs(
    recharts.RadialBarChart,
    {
      margin: { top: 0, bottom: 0, left: 0, right: 0 },
      innerRadius: "10%",
      outerRadius: "100%",
      cx: "50%",
      cy: "50%",
      barSize,
      startAngle,
      endAngle,
      data: dataWithResolvedColor,
      ...radialBarChartProps,
      children: [
        /* @__PURE__ */ jsxRuntime.jsx(
          recharts.RadialBar,
          {
            label: withLabels ? {
              position: "insideStart",
              fill: "var(--mantine-color-white)",
              fontSize: 12
            } : void 0,
            background: withBackground ? { fill: "var(--chart-empty-background)" } : void 0,
            dataKey,
            isAnimationActive: false,
            ...radialBarProps
          }
        ),
        withLegend && /* @__PURE__ */ jsxRuntime.jsx(
          recharts.Legend,
          {
            verticalAlign: "bottom",
            content: (payload) => /* @__PURE__ */ jsxRuntime.jsx(
              ChartLegend.ChartLegend,
              {
                payload: payload.payload?.map((item) => ({
                  ...item,
                  dataKey: item.payload?.name
                })),
                onHighlight: setHighlightedArea,
                legendPosition: legendProps?.verticalAlign || "bottom",
                classNames: resolvedClassNames,
                styles: resolvedStyles,
                centered: true
              }
            ),
            ...legendProps
          }
        ),
        withTooltip && /* @__PURE__ */ jsxRuntime.jsx(
          recharts.Tooltip,
          {
            animationDuration: 0,
            isAnimationActive: false,
            cursor: { stroke: "var(--chart-cursor-color)" },
            content: ({ payload }) => /* @__PURE__ */ jsxRuntime.jsxs(core.Paper, { ...getStyles("tooltip"), children: [
              /* @__PURE__ */ jsxRuntime.jsxs(core.Group, { gap: "sm", children: [
                /* @__PURE__ */ jsxRuntime.jsx(core.ColorSwatch, { color: payload?.[0]?.payload.fill, size: 12, withShadow: false }),
                /* @__PURE__ */ jsxRuntime.jsx("span", { children: payload?.[0]?.payload.name })
              ] }),
              /* @__PURE__ */ jsxRuntime.jsx("span", { children: payload?.[0]?.payload[dataKey] })
            ] }),
            ...tooltipProps
          }
        )
      ]
    }
  ) }) });
});
RadialBarChart.displayName = "@mantine/core/RadialBarChart";
RadialBarChart.classes = RadialBarChart_module;

exports.RadialBarChart = RadialBarChart;
